### PR TITLE
UnoCSS Plugin

### PR DESCRIPTION
**Pull request by kwaa**

### CODE REVIEW

**Comment by kwaa** [comment]
Hmm... just figured out that extending UserConfig might be buggy, I'll fix it tomorrow

**Comment by kwaa** [comment]
I think it could work well now.

**Comment by oscarotero** [comment]
Hi.
I was reviewing the plugin but not sure this is a good approach.

The WindiCSS plugin can export the CSS code to a file and only if the filename is missing, inlines the code in the html pages. This plugin only does the later.

You're including raw CSS code in the plugin. I don't understand why you do that.

**Comment by kwaa** [comment]
> The WindiCSS plugin can export the CSS code to a file and only if the filename is missing, inlines the code in the html pages. This plugin only does the later.

Ah, missed this one. It's slightly more complicated...

> You're including raw CSS code in the plugin. I don't understand why you do that.

UnoCSS puts the CSS reset in a separate `@unocss/reset` package instead of being generated by the generator. And Deno doesn't currently support CSS Modules, so I can't import them, only inline.

**Comment by oscarotero** [comment]
> UnoCSS puts the CSS reset in a separate @unocss/reset package instead of being generated by the generator. And Deno doesn't currently support CSS Modules, so I can't import them, only inline.

Okay but it can be loaded as a remote file or with a fetch.

**Comment by kwaa** [comment]
> Okay but it can be loaded as a remote file or with a fetch.

If remoteFile is used, the user needs to import them manually. If you want, I can change it to fetch

btw: which CDN to get it from? esm.sh / cdn.jsdelivr.net / unpkg.com / etc

**Comment by oscarotero** [comment]
Okay, let's go with fetch then.

You can use the `read` function (https://github.com/lumeland/lume/blob/master/core/utils.ts#L460) so the request is cached the first time, so the plugin can work offline.

I'd use `unpkg.com`.

**Comment by kwaa** [comment]
has been replaced with fetch, I'll take a look at the `cssFiles` option in the following days.

**Comment by kwaa** [comment]
@oscarotero Ran into a bit of a problem: UnoCSS doesn't seem to have a `StyleSheet` class, and I'm not quite sure how to merge the CSS from all the pages.

File: unocss/mod.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      // Insert a <link> tag for each page
+      site.process([".html"], (page) => {
+        const link = page.document!.createElement("link");
+        link.setAttribute("rel", "stylesheet");
+        link.setAttribute("href", site.url(options.cssFile as string));
+        page.document?.head?.appendChild(link);
+      });
+
+      // Generate the stylesheets for all pages
+      site.addEventListener("afterRender", async () => {
```
Comment (oscarotero) [inline unocss/mod.ts:76]: You can use `site.processAll([".html"]`, cb)` to get all html pages. More info: https://lume.land/docs/core/processors/#process-all-pages

File: unocss/mod.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        await Promise.all(
+          pages.map(async (page) =>
+            fileClassMap.set(
+              page.src,
+              await uno.generate(
+                page.document?.documentElement?.innerHTML ?? "",
+              ).then((res) => res.matched),
+            )
+          ),
+        );
```
Comment (oscarotero) [inline unocss/mod.ts:92]: Looking at the documentation, seems like you can define the html code to scan in the `content.inline` option:
https://unocss.dev/guide/extracting#extracting-from-inline-text

**Review by oscarotero — CHANGES_REQUESTED** [review]

**Comment by oscarotero** [comment]
Thanks for your work. I have reviewed your code and let you a couple of comments. Let me know if you have any question :)

File: unocss/mod.ts (hunk 2 above)
Comment (kwaa) [reply unocss/mod.ts:89]: > Looking at the documentation, seems like you can define the html code to scan in the `content.inline` option: https://unocss.dev/guide/extracting#extracting-from-inline-text

I'm not sure I understand what you mean.

Comment (oscarotero) [reply unocss/mod.ts:89]: You are generating a `fileClassMap` running `uno.generate()` for every page and then run again to merge the stylesheets.
[According to the documentation,](https://unocss.dev/guide/extracting#extracting-from-inline-text) it's possible to simply pass an array of all content:

```ts
const content = pages.map((page) => page.content);

const unoCssConfig = {
  ...otherConfig,
  content,
}
```

Comment (kwaa) [reply unocss/mod.ts:89]: and I just tried it and it seems that `@unocss/core` doesn't read `config.content.inline`

Comment (oscarotero) [reply unocss/mod.ts:89]: Ah, ok.
Let's keep it as is, then.

**Comment by oscarotero** [comment]
Thanks!

**Comment by kwaa** [comment]
> Thanks!

Ok, I'm going to make a few more minor changes and will submit another PR after that.
Also will migrating it to Lume v2 allow me to PR? Just want some more commits.

**Pull request merged**